pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java

```diff
@@ -299,16 +277,40 @@ private List<TableConfig> getListTableConfigs(List<String> tableNames) {
     return allTableConfigList;
   }
 
+  private String selectInstanceId(List<String> instanceIds) {
+    if (instanceIds.isEmpty()) {
+      return QueryException.BROKER_RESOURCE_MISSING_ERROR.toString();
+    }
+
+    instanceIds.retainAll(_pinotHelixResourceManager.getOnlineInstanceList());
+    if (instanceIds.isEmpty()) {
+      return QueryException.BROKER_INSTANCE_MISSING_ERROR.toString();
+    }
+
+    // Send query to a random broker.
+    return instanceIds.get(RANDOM.nextInt(instanceIds.size()));
+  }
+
+  private List<String> findCommonBrokerInstance(Set<String> brokerTenants) {
+    Set<String> commonInstances = null;
+    for (String brokerTenant : brokerTenants) {
+      Set<String> instances = _pinotHelixResourceManager.getAllInstancesForBrokerTenant(brokerTenant);
+      if (commonInstances == null) {
+        commonInstances = instances;
+      } else {
+        commonInstances.retainAll(instances);
+      }
+    }
+    return commonInstances == null ? Collections.emptyList() : new ArrayList<>(commonInstances);
+  }
+
   // return the brokerTenant if all table configs point to the same broker, else returns null
-  private String getCommonBrokerTenant(List<TableConfig> tableConfigList) {
-    Set<String> tableBrokers = new HashSet<>();
+  private Set<String> getCommonBrokerTenants(List<TableConfig> tableConfigList) {
```

Comment:
This is no longer the common broker tenants, but all broker tenants

Reply:
will rename